src/Microsoft.Identity.Web.TokenAcquisition/TokenAcquisition.cs

```diff
@@ -585,10 +608,82 @@ private bool IsInvalidClientCertificateOrSignedAssertionError(MsalServiceExcepti
+        /// <returns>The application object used to request a token with managed identity.</returns>
+        internal async Task<IManagedIdentityApplication> GetOrBuildManagedIdentityApplication(
+            MergedOptions mergedOptions, ManagedIdentityOptions managedIdentityOptions)
+        {
+            string key = GetCacheKeyForManagedId(managedIdentityOptions);
+
+            // Check if the application is already built, if so return it without grabbing the lock
+            if (_managedIdentityApplicationsByClientId.TryGetValue(key, out IManagedIdentityApplication? application))
+            {
+                return application;
+            }
+
+            // Lock the potential write of the dictionary to prevent multiple threads from creating the same application.
+            await _managedIdSemaphore.WaitAsync();
+            try
+            {
+                // Check if the application is already built (could happen between previous check and obtaining the key)
+                if (_managedIdentityApplicationsByClientId.TryGetValue(key, out application))
+                {
+                    return application;
+                }
+
+                // Set managedIdentityId to the correct value for either system or user assigned
+                ManagedIdentityId managedIdentityId;
+                if (key == SystemAssignedManagedIdentityKey)
+                {
+                    managedIdentityId = ManagedIdentityId.SystemAssigned;
+                }
+                else
+                {
+                    managedIdentityId = ManagedIdentityId.WithUserAssignedClientId(key);
+                }
+    
+                // Build the application
+                application = BuildManagedIdentityApplication(
+                    managedIdentityId,
+                    mergedOptions.ConfidentialClientApplicationOptions.EnablePiiLogging
+                );
 
+                // Add the application to the cache
+                _managedIdentityApplicationsByClientId.TryAdd(key, application);
+            }
+            finally
+            {
+                // Now that the dictionary is updated, release the semaphore
+                _managedIdSemaphore.Release();
+            }
             return application;
         }
 
+        /// <summary>
+        /// Gets the key value for the Managed Identity cache, the default key for system-assigned identity is used if there is
+        /// no clientId for a user-assigned identity specified. The method is internal rather than private for testing purposes.
+        /// </summary>
+        /// <param name="managedIdOptions">Holds the clientId for managed identity if none is present</param>
+        /// <returns>A key value for the Managed Identity cache</returns>
+        internal static string GetCacheKeyForManagedId(ManagedIdentityOptions managedIdOptions) 
+        {
+            if (managedIdOptions.UserAssignedClientId.IsNullOrEmpty())
```

Comment:
This is going to be interesting....

So let's imagine that we want to add support for UAMI + object ID. You'd go to Abstractions and update `ManagedIdentityOptions` and add "UserAssignedObjectId" and release Abstractions. Then a few days later, release Id.Web with new dependency.

However, if a customer mismatches the version of Id.Web with the version of Abstractions, then they could get into a state where: 

- they set ManagedIdentityOptions's ObjectId  = "bla"
- older Id.Web doesn't know about "ObjectId" so it interpreset this as being System Assigned MSI (SAMI)
- if a SAMI or a single UAMI (even different the one specified) exists, a token for it will be returned. But not the expected one.

My recommendation would be to update ManagedIdentityOptions to: 


```cshapr
public class ManagedIdentityOptions
    
 [DefaultValue(null)]
 public string? UserAssignedClientId { get; set; }
  
 public bool IsSystemAssigned
 { 
      get{ return UserAssignedClientId != null; }
 }
```

Reply:
I definitely agree and this is a neat way to handle the issue so I made [an ADO item for it](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/2816904) since it requires a release in abstractions before being implemented, it is out of scope for this PR but should be done.